reactive/calico.py

```diff
@@ -188,11 +192,17 @@ def reconfigure_calico_pool():
     remove_state('calico.pool.configured')
 
 
-@when('etcd.available', 'cni.is-worker')
+@when('etcd.available')
+# the interface only sets is-master or is-worker when the config is set, but
+# it doesn't indicate it with a is-configured or something. So we have to
+# look for either of them to verify cni is configured
+@when_any('cni.is-master', 'cni.is-worker')
```

Comment:
This feels weird to me.  If I parse your comment correctly, it's that there may be some period of time where `cni.connected` is set, yet `cni.is-[master|worker]` is not due to the provider not having set any config. If that's true, the real issue is that you don't want to proceed in this handler until that config is present.

It seems plausible this could happen when `joined` fires and there is no config yet (hence, `connected`, but no `[master|worker`] flags), and then `changed` fires and we have something.

This feels like we're trying to work around some hook timing, and that never ends well.  Would you consider removing the `@when_any`, adding `@when(cni.connected)`, and then do something like this in the handler:

```python
cni_config = cni.get_config()
if not cni_config.get('kubeconfig_path'):
    return
else:
  ...
```

That makes it more obvious to me that while we may be connected, we don't want to proceed until the provider has given us some config.

Reply:
Your read is correct. Connected means that we are related, but there may or may not be data. If there is data then is-master/is-worker is set. The problem with changing this to just bail on no data is that we will have to wait for the next set of hooks to run to hit this code again instead of being reactive to data showing up. While I'd like to set is-master/is-worker AND another flag at the same time, this is a reasonable compromise. Maybe I should just change the interface to set a flag?

Reply:
If we bail on no data, then that means we didn't have the data we needed in the first place.  Your method indirectly ensures config data is present, but to me, that's more confusing than explicitly checking for what we need and bailing if it's not there.

As for waiting on the next set of hooks -- if there's no config, then we bail.  As soon as the provider sets config, we'll get a `changed` hook that will run this handler.  I don't think we'll have a noticeable penalty in that scenario.

That said, a new flag that is set when config is present would solve this cleanly too.  I guess what I don't like about the current impl is that this is really about `cni_config['kubeconfig_path']`, yet we're masking that by a set of flags that indirectly makes sure that config is present.

Also the enormous comment splitting up those decorators bugs the crap out of me.